File: src/frontend/invoice-web/src/app/shared/pagination/paged-response.ts
/**
 * Generic envelope returned by the server-side paginated listing endpoints
 * (`GET /api/products/paged`, `GET /api/invoices/paged`). Both `Inventory.Api`
 * and `Billing.Api` expose the same JSON shape from their own independent
 * `PagedResponse<T>` type (see `docs/technical-details.md`, "Paginação
 * server-side"), so a single generic frontend interface covers both.
 */
export interface PagedResponse<T> {
  readonly items: T[];
  readonly pageNumber: number;
  readonly pageSize: number;
  readonly totalCount: number;
  readonly totalPages: number;
  readonly hasPreviousPage: boolean;
  readonly hasNextPage: boolean;
}

/** Page sizes offered by the pagination control, in this fixed order. */
export const PAGE_SIZE_OPTIONS = [5, 10, 25, 50] as const;

/** Default page number/size used when none is provided or valid in the route. */
export const DEFAULT_PAGE_NUMBER = 1;
export const DEFAULT_PAGE_SIZE = 5;

File: src/backend/Billing.Api/Common/PagedResponse.cs
namespace Billing.Api.Common;

/// <summary>
/// Generic envelope for a server-side paginated listing. Kept local to
/// Billing.Api (no shared library between microservices) since it is a
/// small, self-contained DTO with no dependency on domain types.
/// </summary>
/// <typeparam name="T">Type of the items returned in the current page.</typeparam>
public record PagedResponse<T>(
    IReadOnlyList<T> Items,
    int PageNumber,
    int PageSize,
    int TotalCount,
    int TotalPages,
    bool HasPreviousPage,
    bool HasNextPage)
{
    /// <summary>
    /// Builds a <see cref="PagedResponse{T}"/> computing <see cref="TotalPages"/>,
    /// <see cref="HasPreviousPage"/> and <see cref="HasNextPage"/> from the given
    /// page number, page size and total count.
    /// </summary>
    public static PagedResponse<T> Create(IReadOnlyList<T> items, int pageNumber, int pageSize, int totalCount)
    {
        var totalPages = totalCount == 0 ? 0 : (int)Math.Ceiling(totalCount / (double)pageSize);

        return new PagedResponse<T>(
            items,
            pageNumber,
            pageSize,
            totalCount,
            totalPages,
            HasPreviousPage: pageNumber > 1,
            HasNextPage: pageNumber < totalPages);
    }
}
